pkg/kubelet/kuberuntime/kuberuntime_container_linux.go

```diff
@@ -99,21 +102,17 @@ func (m *kubeGenericRuntimeManager) generateLinuxContainerResources(pod *v1.Pod,
 
 	lcr.HugepageLimits = GetHugepageLimitsFromResources(container.Resources)
 
-	if utilfeature.DefaultFeatureGate.Enabled(kubefeatures.NodeSwap) {
+	if swapConfigurationHelper := newSwapConfigurationHelper(*m.machineInfo); utilfeature.DefaultFeatureGate.Enabled(kubefeatures.NodeSwap) {
 		// NOTE(ehashman): Behaviour is defined in the opencontainers runtime spec:
 		// https://github.com/opencontainers/runtime-spec/blob/1c3f411f041711bbeecf35ff7e93461ea6789220/config-linux.md#memory
 		switch m.memorySwapBehavior {
-		case kubelettypes.UnlimitedSwap:
-			// -1 = unlimited swap
-			lcr.MemorySwapLimitInBytes = -1
 		case kubelettypes.LimitedSwap:
-			fallthrough
+			swapConfigurationHelper.ConfigureLimitedSwap(lcr, pod, container)
 		default:
-			// memorySwapLimit = total permitted memory+swap; if equal to memory limit, => 0 swap above memory limit
-			// Some swapping is still possible.
-			// Note that if memory limit is 0, memory swap limit is ignored.
-			lcr.MemorySwapLimitInBytes = lcr.MemoryLimitInBytes
+			swapConfigurationHelper.ConfigureUnlimitedSwap(lcr)
 		}
+	} else {
+		swapConfigurationHelper.ConfigureNoSwap(lcr)
```

Comment:
it is what we have today in alpha? If feature gate is disabled, do we configure no swap? I thought we do not make any configuration in this case

Reply:
You're right, today we don't configure anything if the feature gate is disabled.

However, the question is if we want to ensure that swap is not accessible if the feature gate is off, or do we want to let the CRI decide what to do in this case. I know that in cgroups, the default is using `max`, which is equivalent to `UnlimitedSwap`. Maybe most CRIs default to `NoSwap`, but IMHO it's better to actively ensure no swap is accessible if the feature gate is off.

However, if you don't agree I can get rid of this logic.

WDYT?

Reply:
I am trying to understand whether this will break any customer who don't use the feature gate at all. 

Let's say somebody was running with `failSwapOn: false` before. 

If they never enabled nor disabled the feature gate, they would have the default container runtime's behavior today. Once upgraded to 1.28, they will have the feature gate enabled by default (default of feature gate is `true`). And they will get the unlimited behavior. So instead of container runtime's default behavior we will get the unlimited set explicitly.

If they explicitly enabled the feature gate, they will get the same behavior with `beta1` as was configured with `alpha`. So this is good - no surprises.

If they explicitly disabled the feature gate in `alpha`, after the upgrade they will get explicit zero instead of Container Runtime. I think Zero is a default behavior for Containerd at least latest versions.

So the actual concern is not this code branch. More concerning the new default with no feature gate set. Maybe we need to rethink the default behavior from being unlimited to being whatever Container Runtime default is?

Reply:
If we cannot find a quick answer here, perhaps what we can do:

- Add a note in this branch suggesting the change of behavior when featuer gate is disabled explicitly. Relnote it as well
- Turn the feature gate off by default. Deal with the upgrade scenario in `beta2`

Reply:
Yeah, let's keep it off by default and revisit.

Reply:
@SergeyKanzhelev Thanks for elaborating.

> Yeah, let's keep it off by default and revisit.

Yup, I agree.
Made the change :+1: PTAL